File: TCP  IP Server Client Example Part 6 (Chat Application)/[VB]-TCP  IP Server Client Example Part 6 (Chat Application)/VB/ServerClientChat1.3/FrmServer.vb
Imports NetComm
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Imports ServerClientChat1._3.Classes
Imports ServerClientChat1._3.Forms

Namespace ServerClientChat1._3
	Partial Public Class FrmServer
		Inherits Form

		Private Shared ReadOnly OurPort As Integer = 3000 ' The Server's Port
		Private server As NetComm.Host = New Host(OurPort) ' Listens on port 3330.
		Private pip As New PublicIP() ' Class to get Server's Public IP Address
		Private AzureDB As New AzureSQL() ' Class for communicating with Azure Database

		Private TotalConnectedUsers As Integer = 0 ' In production you may want to use longs instead of ints.
		Private TotalRegisteredUsers As Integer = 0 ' In production you may want to use longs instead of ints.
		Private TotalLoggedInUsers As Integer = 0 ' In production you may want to use longs instead of ints.

		Private Delegate Sub UpdateConnectedUsersDelegate(ByVal s As String)
		Public Delegate Sub UpdateLogDelegate(ByVal txt As String)
		Public Delegate Sub IPAddressDelegate(ByVal ip As String)
		Public Delegate Sub UpdateTotalLoggedInUsersDelegate(ByVal number As String)
		Public Delegate Sub UpdateTotalRegisteredUsersDelegate(ByVal number As Integer)

		Public Sub New()
			InitializeComponent()
		End Sub

		''' <summary>
		''' Start The server and Setup initial variables and connections
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub FrmServer_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			' Get Azure Server Name, Table Name, User name and Password

			Dim AzureLogin As New FrmAzureDatabaseLogin()
			Dim dr As DialogResult = AzureLogin.ShowDialog()
			If dr = System.Windows.Forms.DialogResult.OK Then

				AddHandler AzureDB.SQLInfoMessage, AddressOf AzureDB_SQLInfoMessage ' Get any messages from the SQL Database
				AddHandler AzureDB.SQLStateChanged, AddressOf AzureDB_SQLStateChanged ' Keep track of Database Connection Changes

				AzureDB.ConnectionString = "Server=tcp:" & AzureLogin.ServerName & ";Database=ChatApp;User ID=" & AzureLogin.UserName & ";Password=" & AzureLogin.Password & ";Trusted_Connection=False;Encrypt=True;Connection Timeout=30;"
				If AzureDB.Connect() Then
					' Display our Public IP Address and the port we are listening to.
					pip = New PublicIP()
					AddHandler pip.PublicIPKnown, AddressOf pip_PublicIPKnown ' Get's the Public IP Address of the Server (Event)
					AddHandler pip.PublicIPError, AddressOf pip_PublicIPError ' Notifies us of any Errors Getting the Public IP
					pip.GetPublicIpAddress() ' Will Raise the above event
					lblPort.Text = OurPort.ToString()

					' Subscribe to Server Events
					AddHandler server.ConnectionClosed, AddressOf server_ConnectionClosed
					AddHandler server.DataReceived, AddressOf server_DataReceived
					AddHandler server.DataTransferred, AddressOf server_DataTransferred
					AddHandler server.errEncounter, AddressOf server_errEncounter
					AddHandler server.lostConnection, AddressOf server_lostConnection
					AddHandler server.onConnection, AddressOf server_onConnection
					server.StartConnection() ' Don't forget to start your server !!

					' Now get total number of Registered Users.
					TotalRegisteredUsers = AzureDB.GetRegisteredUsers()
					UpdateTotalRegisteredUsers(TotalRegisteredUsers)
				Else
					MessageBox.Show("Azure Database Error: " & AzureDB.LastError)
					lblConnectionState.Image = My.Resources.database_red
				End If

			Else
				MessageBox.Show("The Server cannot run unless connected to the Server's Azure SQL Database " & vbLf & vbCr & "Closing Application")
				Me.Close()
			End If
		End Sub



		Private Sub pip_PublicIPError(ByVal Message As String)
			Log("Public IP: " & Message)
		End Sub

		Private Sub server_onConnection(ByVal id As String)
			Me.TotalConnectedUsers += 1
			UpdateConnectedUser(TotalConnectedUsers.ToString("N0"))

		End Sub

		Private Sub UpdateConnectedUser(ByVal s As String)
			If Me.lblTotalConnectedUsers.InvokeRequired Then
				Dim d As New UpdateConnectedUsersDelegate(AddressOf UpdateConnectedUser)
				Me.lblTotalConnectedUsers.Invoke(d, New Object() { s })
			Else
				Me.lblTotalConnectedUsers.Text = s

			End If
		End Sub

		Private Sub UpdateTotalRegisteredUsers(ByVal TotalRegisteredUsers As Integer)
			If Me.lblTotalRegisteredUsers.InvokeRequired Then
				Dim d As New UpdateTotalRegisteredUsersDelegate(AddressOf UpdateTotalRegisteredUsers)
				Me.lblTotalConnectedUsers.Invoke(d, New Object() { TotalRegisteredUsers })
			Else
				Me.lblTotalRegisteredUsers.Text = TotalRegisteredUsers.ToString("N0")

			End If
		End Sub

		Private Sub server_lostConnection(ByVal id As String)
			UpdateLoggedInUsers(TotalLoggedInUsers.ToString("N0"))
			UpdateConnectedUser(TotalConnectedUsers.ToString("N0"))
		End Sub

		Private Sub server_errEncounter(ByVal ex As Exception)
			Throw New NotImplementedException()
		End Sub

		Private Sub server_DataTransferred(ByVal Sender As String, ByVal Recipient As String, ByVal Data() As Byte)
			Dim message As String = ConvertBytesToString(Data)
			If Recipient = "0" Then ' The Server's ID
				Log("Received Command From Client " & Sender & " for this Server")


				If message.StartsWith("CLOSING") Then
					DoClientClosing(Sender)
				ElseIf message.StartsWith("REGISTRATION") Then
					DoRegistration(Sender, message)
				ElseIf message.StartsWith("LOGIN") Then
					DoLogin(Sender, message)
				ElseIf message.StartsWith("RELOGIN") Then
					DoRelogin(Sender, message)
				Else
					Me.Log("Received Command From Client " & Sender & " for this Server: " & message)
				End If


			Else
				Me.Log("Received Command From Client " & Sender & " for this Server: " & message)
			End If
		End Sub

		Private Sub DoClientClosing(ByVal Sender As String)
			Me.TotalConnectedUsers -= 1
			UpdateConnectedUser(TotalConnectedUsers.ToString("N0"))
			Me.Log("Client " & Sender & " is Closing")
		End Sub

		Private Sub DoRegistration(ByVal Sender As String, ByVal message As String)
			' Message = REGISTRATION:<USERNAME>:<PASSWORD>

			If Not AzureDB.RegisterUser(message.Split(":"c)(1).ToString(), message.Split(":"c)(2).ToString()) Then
				Me.server.SendData(Sender, ConvertStringToBytes("ERROR:That email is already registered."))
				Me.Log("Client " & Sender & "ERROR:That email is already registered.")
			Else
				Dim cID As Integer = AzureDB.GetUsersID(message.Split(":"c)(1).ToString())
				Me.TotalLoggedInUsers += 1 ' Add this user to the logged in user count
				Me.TotalRegisteredUsers += 1
				UpdateTotalRegisteredUsers(TotalRegisteredUsers)
				Me.server.SendData(Sender, ConvertStringToBytes("SUCCESS:Logged On " & cID))
				Me.Log("Client " & Sender & "SUCCESS registered.")
			End If
		End Sub

		Private Sub DoRelogin(ByVal Sender As String, ByVal message As String)
			' Message = RELOGIN:<USERNAME>:<PASSWORD>
			Dim parts() As String = message.Split(":"c)
			If AzureDB.LoginUser(message.Split(":"c)(1).ToString(), message.Split(":"c)(2).ToString()) Then
				' Successfully logged in With Correct User ID
				Me.TotalLoggedInUsers += 1
				UpdateLoggedInUsers(TotalLoggedInUsers.ToString("N0"))

				Me.Log("Client " & parts(1).ToString() & " SUCCESS Logged in.")
			Else
				Me.server.SendData(Sender, ConvertStringToBytes("ERROR:Login Details Not Recognized"))
				Me.Log("Client " & Sender & "ERROR:Login Details Not Recognized")
			End If
		End Sub

		Private Sub DoLogin(ByVal Sender As String, ByVal message As String)
			' Message = LOGIN:<USERNAME>:<PASSWORD>
			Dim parts() As String = message.Split(":"c)
			If AzureDB.RegisterUserExists(parts(1).ToString()) Then

				If AzureDB.LoginUser(message.Split(":"c)(1).ToString(), message.Split(":"c)(2).ToString()) Then
					' Successfully logged in
					Dim newSenderId As Integer = AzureDB.GetUsersID(message.Split(":"c)(1).ToString()) ' Get this user's client ID
					Me.server.SendData(Sender, ConvertStringToBytes("ID:" & newSenderId.ToString())) ' Send the correct ID to the Client
					Me.TotalLoggedInUsers += 1
					UpdateLoggedInUsers(TotalLoggedInUsers.ToString("N0"))

					Me.Log("Client " & parts(1).ToString() & " SUCCESS Logged in.")
				Else
					Me.server.SendData(Sender, ConvertStringToBytes("ERROR:Login Details Not Recognized"))
					Me.Log("Client " & Sender & "ERROR:Login Details Not Recognized")
				End If
			Else
				Me.server.SendData(Sender, ConvertStringToBytes("ERROR:Login Details Not Recognized"))
				Me.Log("Client " & Sender & "ERROR:Login Details Not Recognized")
			End If
		End Sub



		Private Sub server_DataReceived(ByVal ID As String, ByVal Data() As Byte)
			Throw New NotImplementedException()
		End Sub

		Private Sub server_ConnectionClosed()
			Throw New NotImplementedException()
		End Sub

		''' <summary>
		''' The Azure Database Connection has changed.
		''' </summary>
		''' <param name="State"></param>
		Private Sub AzureDB_SQLStateChanged(ByVal State As ConnectionState)
			Select Case State
				Case ConnectionState.Broken
					lblConnectionState.Image = My.Resources.database_red ' Show Database connection is off
					lblConnectionState.ToolTipText = "Connection to Azure DB is " & State.ToString()
					Progress.Enabled = False
				Case ConnectionState.Closed
					lblConnectionState.Image = My.Resources.database_red ' Show Database connection is off
					lblConnectionState.ToolTipText = "Connection to Azure DB is " & State.ToString()
					Progress.Enabled = False
				Case ConnectionState.Connecting
					lblConnectionState.Image = My.Resources.database_blue ' Show Database connection is Connecting
					lblConnectionState.ToolTipText = "Connection to Azure DB is " & State.ToString()
					Progress.Enabled = True
				Case ConnectionState.Executing
					lblConnectionState.Image = My.Resources.database_blue ' Show Database connection is Connecting
					lblConnectionState.ToolTipText = "Connection to Azure DB is " & State.ToString()
					Progress.Enabled = True
				Case ConnectionState.Fetching
					lblConnectionState.Image = My.Resources.database_blue ' Show Database connection is Connecting
					lblConnectionState.ToolTipText = "Connection to Azure DB is " & State.ToString()
					Progress.Enabled = True
				Case ConnectionState.Open
					lblConnectionState.Image = My.Resources.database_green ' Show Database connection is Connected
					lblConnectionState.ToolTipText = "Connection to Azure DB is " & State.ToString()
					Progress.Enabled = True
				Case Else
					lblConnectionState.Image = My.Resources.database_red ' Show Database connection is off
					lblConnectionState.ToolTipText = "Connection to Azure DB is " & State.ToString()
					Progress.Enabled = False
			End Select
		End Sub

		''' <summary>
		''' We have a new MEssage from the Azure Database
		''' </summary>
		''' <param name="Message"></param>
		Private Sub AzureDB_SQLInfoMessage(ByVal Message As String)
			Log(Message)
		End Sub

		''' <summary>
		''' Display Logging Messages to Admin
		''' </summary>
		''' <param name="Message"></param>
		Private Sub Log(ByVal Message As String)
			If Me.rtbConOut.InvokeRequired Then
				Dim d As New UpdateLogDelegate(AddressOf Log)
				Me.rtbConOut.Invoke(d, New Object() { Message })
			Else
				rtbConOut.AppendText(Message & Environment.NewLine)
				rtbConOut.Focus()
			End If
		End Sub

		Private Sub UpdateLoggedInUsers(ByVal p As String)
			If Me.lblTotalLoggedInUsers.InvokeRequired Then
				Dim d As New UpdateTotalLoggedInUsersDelegate(AddressOf UpdateLoggedInUsers)
				Me.lblTotalLoggedInUsers.Invoke(d, New Object() { p })
			Else
				lblTotalLoggedInUsers.Text = p

			End If
		End Sub

		''' <summary>
		''' Sets the Public IP Address Label
		''' </summary>
		''' <param name="PublicIP">
		''' String: The public IP Address that the Server is showing the world
		''' </param>
		Private Sub pip_PublicIPKnown(ByVal PublicIP As String)
			If Me.lblIPAddress.InvokeRequired Then
				Dim d As New IPAddressDelegate(AddressOf pip_PublicIPKnown)
				Me.lblIPAddress.Invoke(d, New Object() { PublicIP })
			Else
				lblIPAddress.Text = PublicIP
			End If
		End Sub

		''' <summary>
		''' Do house maintenance before closing.
		''' Including Notifying all users if possible
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub FrmServer_FormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs) Handles MyBase.FormClosing
			' Unsubscribe to Server Events
			RemoveHandler server.ConnectionClosed, AddressOf server_ConnectionClosed
			RemoveHandler server.DataReceived, AddressOf server_DataReceived
			RemoveHandler server.DataTransferred, AddressOf server_DataTransferred
			RemoveHandler server.errEncounter, AddressOf server_errEncounter
			RemoveHandler server.lostConnection, AddressOf server_lostConnection
			RemoveHandler server.onConnection, AddressOf server_onConnection

			' Unsubscribe to PublicIP Events
			RemoveHandler pip.PublicIPKnown, AddressOf pip_PublicIPKnown
			RemoveHandler pip.PublicIPError, AddressOf pip_PublicIPError

			' Unsubscribe to AzureDB Events
			RemoveHandler AzureDB.SQLInfoMessage, AddressOf AzureDB_SQLInfoMessage
			RemoveHandler AzureDB.SQLStateChanged, AddressOf AzureDB_SQLStateChanged
		End Sub



		''' <summary>
		''' Convert Messages from the Server into English
		''' </summary>
		''' <param name="bytes"></param>
		''' <returns></returns>
		Private Function ConvertBytesToString(ByVal bytes() As Byte) As String
			Return ASCIIEncoding.ASCII.GetString(bytes)
		End Function

		''' <summary>
		''' Convert English messages into Messages for the server
		''' </summary>
		''' <param name="str"></param>
		''' <returns></returns>
		Private Function ConvertStringToBytes(ByVal str As String) As Byte()
			Return ASCIIEncoding.ASCII.GetBytes(str)
		End Function
	End Class
End Namespace

File: TCP  IP Server Client Example Part 6 (Chat Application)/[C#]-TCP  IP Server Client Example Part 6 (Chat Application)/C#/ServerClientChat1.3/FrmServer.cs
using NetComm;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

using ServerClientChat1._3.Classes;
using ServerClientChat1._3.Forms;

namespace ServerClientChat1._3
{
    public partial class FrmServer : Form
    {

        private static readonly int OurPort = 3000;         // The Server's Port
        private NetComm.Host server = new Host(OurPort);    // Listens on port 3330.
        private PublicIP pip = new PublicIP();              // Class to get Server's Public IP Address
        private AzureSQL AzureDB = new AzureSQL();          // Class for communicating with Azure Database
      
        private int TotalConnectedUsers = 0;                // In production you may want to use longs instead of ints.
        private int TotalRegisteredUsers = 0;               // In production you may want to use longs instead of ints.
        private int TotalLoggedInUsers = 0;               // In production you may want to use longs instead of ints.

        private delegate void UpdateConnectedUsersDelegate(string s);
        public delegate void UpdateLogDelegate(string txt);
        public delegate void IPAddressDelegate(string ip);
        public delegate void UpdateTotalLoggedInUsersDelegate(string number);
        public delegate void UpdateTotalRegisteredUsersDelegate(int number);

        public FrmServer()
        {
            InitializeComponent();
           }

        /// <summary>
        /// Start The server and Setup initial variables and connections
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void FrmServer_Load(object sender, EventArgs e)
        {
            // Get Azure Server Name, Table Name, User name and Password

            FrmAzureDatabaseLogin AzureLogin = new FrmAzureDatabaseLogin();
            DialogResult dr = AzureLogin.ShowDialog();
            if (dr == System.Windows.Forms.DialogResult.OK)
            {

                AzureDB.SQLInfoMessage += AzureDB_SQLInfoMessage; // Get any messages from the SQL Database
                AzureDB.SQLStateChanged += AzureDB_SQLStateChanged; // Keep track of Database Connection Changes

                AzureDB.ConnectionString = "Server=tcp:" + AzureLogin.ServerName + ";Database=ChatApp;User ID=" + AzureLogin.UserName + ";Password=" + AzureLogin.Password + ";Trusted_Connection=False;Encrypt=True;Connection Timeout=30;";
                if (AzureDB.Connect())
                {
                    // Display our Public IP Address and the port we are listening to.
                    pip = new PublicIP();
                    pip.PublicIPKnown += pip_PublicIPKnown; // Get's the Public IP Address of the Server (Event)
                    pip.PublicIPError += pip_PublicIPError; // Notifies us of any Errors Getting the Public IP
                    pip.GetPublicIpAddress(); // Will Raise the above event
                    lblPort.Text = OurPort.ToString();

                    // Subscribe to Server Events
                    server.ConnectionClosed += server_ConnectionClosed;
                    server.DataReceived += server_DataReceived;
                    server.DataTransferred += server_DataTransferred;
                    server.errEncounter += server_errEncounter;
                    server.lostConnection += server_lostConnection;
                    server.onConnection += server_onConnection;
                    server.StartConnection(); // Don't forget to start your server !!

                    // Now get total number of Registered Users.
                    TotalRegisteredUsers = AzureDB.GetRegisteredUsers();
                    UpdateTotalRegisteredUsers(TotalRegisteredUsers);
                }
                else
                {
                    MessageBox.Show("Azure Database Error: " + AzureDB.LastError);
                    lblConnectionState.Image = ServerClientChat1._3.Properties.Resources.database_red;
                }

            }
            else
            {
                MessageBox.Show("The Server cannot run unless connected to the Server's Azure SQL Database \n\rClosing Application");
                this.Close();
            }
        }

       

        void pip_PublicIPError(string Message)
        {
            Log("Public IP: " + Message);
        }

        void server_onConnection(string id)
        {
            this.TotalConnectedUsers++;
            UpdateConnectedUser(TotalConnectedUsers.ToString("N0"));

        }

        private void UpdateConnectedUser(string s)
        {
            if (this.lblTotalConnectedUsers.InvokeRequired)
            {
                UpdateConnectedUsersDelegate d = new UpdateConnectedUsersDelegate(UpdateConnectedUser);
                this.lblTotalConnectedUsers.Invoke(d, new object[] { s });
            }
            else
            {
                this.lblTotalConnectedUsers.Text = s;

            }
        }

        private void UpdateTotalRegisteredUsers(int TotalRegisteredUsers)
        {
            if (this.lblTotalRegisteredUsers.InvokeRequired)
            {
                UpdateTotalRegisteredUsersDelegate d = new UpdateTotalRegisteredUsersDelegate(UpdateTotalRegisteredUsers);
                this.lblTotalConnectedUsers.Invoke(d, new object[] { TotalRegisteredUsers });
            }
            else
            {
                this.lblTotalRegisteredUsers.Text = TotalRegisteredUsers.ToString("N0");

            }
        }

        void server_lostConnection(string id)
        {
            UpdateLoggedInUsers(TotalLoggedInUsers.ToString("N0"));
            UpdateConnectedUser(TotalConnectedUsers.ToString("N0"));
        }

        void server_errEncounter(Exception ex)
        {
            throw new NotImplementedException();
        }

        void server_DataTransferred(string Sender, string Recipient, byte[] Data)
        {
            string message = ConvertBytesToString(Data);
            if (Recipient == "0") // The Server's ID
            {
                Log("Received Command From Client " + Sender + " for this Server");


                if (message.StartsWith("CLOSING"))
                {
                    DoClientClosing(Sender);
                }
                else if (message.StartsWith("REGISTRATION"))
                {
                    DoRegistration(Sender, message);
                }
                else if (message.StartsWith("LOGIN"))
                {
                    DoLogin(Sender, message);
                }
                else if (message.StartsWith("RELOGIN"))
                {
                    DoRelogin(Sender, message);
                }
                else
                {
                    this.Log("Received Command From Client " + Sender + " for this Server: " + message);
                }


            }
            else
            {
                this.Log("Received Command From Client " + Sender + " for this Server: " + message);
            }
        }

        private void DoClientClosing(string Sender)
        {
            this.TotalConnectedUsers--;
            UpdateConnectedUser(TotalConnectedUsers.ToString("N0"));
            this.Log("Client " + Sender + " is Closing");
        }

        private void DoRegistration(string Sender, string message)
        {
            // Message = REGISTRATION:<USERNAME>:<PASSWORD>

            if (!AzureDB.RegisterUser(message.Split(':')[1].ToString(), message.Split(':')[2].ToString()))
            {
                this.server.SendData(Sender, ConvertStringToBytes("ERROR:That email is already registered."));
                this.Log("Client " + Sender + "ERROR:That email is already registered.");
            }
            else
            {
                int cID = AzureDB.GetUsersID(message.Split(':')[1].ToString());
                this.TotalLoggedInUsers++; // Add this user to the logged in user count
                this.TotalRegisteredUsers++;
                UpdateTotalRegisteredUsers(TotalRegisteredUsers);
                this.server.SendData(Sender, ConvertStringToBytes("SUCCESS:Logged On " + cID));
                this.Log("Client " + Sender + "SUCCESS registered.");
            }
        }

        private void DoRelogin(string Sender, string message)
        {
            // Message = RELOGIN:<USERNAME>:<PASSWORD>
            string[] parts = message.Split(':');
            if (AzureDB.LoginUser(message.Split(':')[1].ToString(), message.Split(':')[2].ToString()))
            {
                // Successfully logged in With Correct User ID
                this.TotalLoggedInUsers++;
                UpdateLoggedInUsers(TotalLoggedInUsers.ToString("N0"));

                this.Log("Client " + parts[1].ToString() + " SUCCESS Logged in.");
            }
            else
            {
                this.server.SendData(Sender, ConvertStringToBytes("ERROR:Login Details Not Recognized"));
                this.Log("Client " + Sender + "ERROR:Login Details Not Recognized");
            }
        }

        private void DoLogin(string Sender, string message)
        {
            // Message = LOGIN:<USERNAME>:<PASSWORD>
            string[] parts = message.Split(':');
            if (AzureDB.RegisterUserExists(parts[1].ToString()))
            {

                if (AzureDB.LoginUser(message.Split(':')[1].ToString(), message.Split(':')[2].ToString()))
                {
                    // Successfully logged in
                    int newSenderId = AzureDB.GetUsersID(message.Split(':')[1].ToString()); // Get this user's client ID
                    this.server.SendData(Sender, ConvertStringToBytes("ID:" + newSenderId.ToString())); // Send the correct ID to the Client
                    this.TotalLoggedInUsers++;
                    UpdateLoggedInUsers(TotalLoggedInUsers.ToString("N0"));

                    this.Log("Client " + parts[1].ToString() + " SUCCESS Logged in.");
                }
                else
                {
                    this.server.SendData(Sender, ConvertStringToBytes("ERROR:Login Details Not Recognized"));
                    this.Log("Client " + Sender + "ERROR:Login Details Not Recognized");
                }
            }
            else
            {
                this.server.SendData(Sender, ConvertStringToBytes("ERROR:Login Details Not Recognized"));
                this.Log("Client " + Sender + "ERROR:Login Details Not Recognized");
            }
        }

       

        private void server_DataReceived(string ID, byte[] Data)
        {
            throw new NotImplementedException();
        }

        void server_ConnectionClosed()
        {
            throw new NotImplementedException();
        }

        /// <summary>
        /// The Azure Database Connection has changed.
        /// </summary>
        /// <param name="State"></param>
        void AzureDB_SQLStateChanged(ConnectionState State)
        {
            switch (State)
            {
                case ConnectionState.Broken:
                    lblConnectionState.Image = ServerClientChat1._3.Properties.Resources.database_red; // Show Database connection is off
                    lblConnectionState.ToolTipText = "Connection to Azure DB is " + State.ToString();
                    Progress.Enabled = false;
                    break;
                case ConnectionState.Closed:
                    lblConnectionState.Image = ServerClientChat1._3.Properties.Resources.database_red; // Show Database connection is off
                    lblConnectionState.ToolTipText = "Connection to Azure DB is " + State.ToString();
                    Progress.Enabled = false;
                    break;
                case ConnectionState.Connecting:
                    lblConnectionState.Image = ServerClientChat1._3.Properties.Resources.database_blue; // Show Database connection is Connecting
                    lblConnectionState.ToolTipText = "Connection to Azure DB is " + State.ToString();
                    Progress.Enabled = true;
                    break;
                case ConnectionState.Executing:
                    lblConnectionState.Image = ServerClientChat1._3.Properties.Resources.database_blue; // Show Database connection is Connecting
                    lblConnectionState.ToolTipText = "Connection to Azure DB is " + State.ToString();
                    Progress.Enabled = true;
                    break;
                case ConnectionState.Fetching:
                    lblConnectionState.Image = ServerClientChat1._3.Properties.Resources.database_blue; // Show Database connection is Connecting
                    lblConnectionState.ToolTipText = "Connection to Azure DB is " + State.ToString();
                    Progress.Enabled = true;
                    break;
                case ConnectionState.Open:
                    lblConnectionState.Image = ServerClientChat1._3.Properties.Resources.database_green; // Show Database connection is Connected
                    lblConnectionState.ToolTipText = "Connection to Azure DB is " + State.ToString();
                    Progress.Enabled = true;
                    break;
                default:
                    lblConnectionState.Image = ServerClientChat1._3.Properties.Resources.database_red; // Show Database connection is off
                    lblConnectionState.ToolTipText = "Connection to Azure DB is " + State.ToString();
                    Progress.Enabled = false;
                    break;
            }
        }

        /// <summary>
        /// We have a new MEssage from the Azure Database
        /// </summary>
        /// <param name="Message"></param>
        void AzureDB_SQLInfoMessage(string Message)
        {
            Log(Message);
        }

        /// <summary>
        /// Display Logging Messages to Admin
        /// </summary>
        /// <param name="Message"></param>
        private void Log(string Message)
        {
            if (this.rtbConOut.InvokeRequired)
            {
                UpdateLogDelegate d = new UpdateLogDelegate(Log);
                this.rtbConOut.Invoke(d, new object[] { Message });
            }
            else
            {
                rtbConOut.AppendText(Message + Environment.NewLine);
                rtbConOut.Focus();
            }
        }

        private void UpdateLoggedInUsers(string p)
        {
            if (this.lblTotalLoggedInUsers.InvokeRequired)
            {
                UpdateTotalLoggedInUsersDelegate d = new UpdateTotalLoggedInUsersDelegate(UpdateLoggedInUsers);
                this.lblTotalLoggedInUsers.Invoke(d, new object[] { p });
            }
            else
            {
                lblTotalLoggedInUsers.Text = p;
              
            }
        }

        /// <summary>
        /// Sets the Public IP Address Label
        /// </summary>
        /// <param name="PublicIP">
        /// String: The public IP Address that the Server is showing the world
        /// </param>
        void pip_PublicIPKnown(string PublicIP)
        {
            if (this.lblIPAddress.InvokeRequired)
            {
                IPAddressDelegate d = new IPAddressDelegate(pip_PublicIPKnown);
                this.lblIPAddress.Invoke(d, new object[] { PublicIP });
            }
            else
            {
                lblIPAddress.Text = PublicIP;
            }
        }

        /// <summary>
        /// Do house maintenance before closing.
        /// Including Notifying all users if possible
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void FrmServer_FormClosing(object sender, FormClosingEventArgs e)
        {
            // Unsubscribe to Server Events
            server.ConnectionClosed -= server_ConnectionClosed;
            server.DataReceived -= server_DataReceived;
            server.DataTransferred -= server_DataTransferred;
            server.errEncounter -= server_errEncounter;
            server.lostConnection -= server_lostConnection;
            server.onConnection -= server_onConnection;

            // Unsubscribe to PublicIP Events
            pip.PublicIPKnown -= pip_PublicIPKnown;
            pip.PublicIPError -= pip_PublicIPError;

            // Unsubscribe to AzureDB Events
            AzureDB.SQLInfoMessage -= AzureDB_SQLInfoMessage;
            AzureDB.SQLStateChanged -= AzureDB_SQLStateChanged;
        }

             

        /// <summary>
        /// Convert Messages from the Server into English
        /// </summary>
        /// <param name="bytes"></param>
        /// <returns></returns>
        string ConvertBytesToString(byte[] bytes)
        {
            return ASCIIEncoding.ASCII.GetString(bytes);
        }

        /// <summary>
        /// Convert English messages into Messages for the server
        /// </summary>
        /// <param name="str"></param>
        /// <returns></returns>
        byte[] ConvertStringToBytes(string str)
        {
            return ASCIIEncoding.ASCII.GetBytes(str);
        }
    }
}
